### PR TITLE
ENH: performance improvements, single file matching, error message improvements [GEAR-362]

### PR DESCRIPTION
**Pull request by kalebdfischer**

### CODE REVIEW

File: transfer_log.py [hunk 1]
```diff
@@ -461,52 +430,303 @@ def load_metadata_table(self):
...
+            id_str = f'{self.config.join}.id'
+            container_id = row.row_dict.get(id_str)
+            path_dict[container_id] = res_path
+        self.resolver_path_dict = path_dict
+        return path_dict
+
+    def get_record_df(self, df):
+
+        # Squash dataframe into one row per set of match column values
+        df = df.groupby(self.match_cols)['tl_index'].apply(list).reset_index()
```
Comment (NPann) [inline transfer_log.py:525]: nice split-apply-combine

File: transfer_log.py [hunk 2]
```diff
@@ -461,52 +430,303 @@ def load_metadata_table(self):
...
+        self.match_df['records_metadata'].fillna(0, inplace=True)
+        self.match_df['records_flywheel'].fillna(0, inplace=True)
+
+        # select rows where transfer log and flywheel match
+        both_df = self.match_df[self.match_df['_merge'] == 'both']
+        # Get list of container IDs with transfer log matches
+        self.matched_containers = list(set([
+            x for array in both_df['tl_index_flywheel'] for x in array
+        ]))
+        return self.match_df
```
Comment (NPann) [inline transfer_log.py:721]: Sweet. so much cleaner and robust :)

File: transfer_log.py [hunk 3]
```diff
@@ -461,52 +430,303 @@ def load_metadata_table(self):
...
+        column_list = ['flywheel_id', 'transfer_log_rows'] + column_list
+        column_list = column_list + ['error', 'matching_fw_ids', 'path']
+        error_df = error_df[column_list]
+
+        return error_df
+
+    @staticmethod
+    def count_df_errors(df):
+        """
+        Counts the number of unique flywheel container IDs and Transfer Log
```
Comment (joshicola) [inline transfer_log.py:677]: Admittedly nitpicky:
Are we going with the succinct short first line of the docstring (w/o hanging indent) and giving a fuller descriptions (if needed) in a follow-up?

https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html

File: transfer_log.py [hunk 4]
```diff
@@ -1,28 +1,29 @@
...
-import flywheel
-import logging
 import json
-import itertools
+import logging
 import os
-import pandas as pd
 import re
+
+import backoff
```
Comment (joshicola) [inline transfer_log.py:15]: "backoff" is imported twice.

"isort" works nice to organize and eliminate dups.

File: transfer_log.py [hunk 5]
```diff
@@ -155,41 +156,41 @@ def load_transfer_log(metadata_path, config):
     return raw_metadata
 
 
-def load_flywheel_records(client, config, project_id):
-    """Load a dictionary with indexes to easily query the project
+@backoff.on_exception(backoff.expo, flywheel.rest.ApiException,
```
Comment (joshicola) [inline transfer_log.py:159]: Nice little decorator.... I am going to use this one!

File: transfer_log.py [hunk 6]
```diff
@@ -155,41 +156,41 @@ def load_transfer_log(metadata_path, config):
...
+                      max_time=300, giveup=utils.false_if_status_gte_500)
+def get_view_from_config(fw_client, config):
+    """
+    Constructs and returns a DataView according to config's specification
 
     Args:
-        client (Client): The flywheel sdk client
-        config (Config): The config object
-        project_id (str): The id of the project to which to migrate metadata
+        fw_client (flywheel.Client): an instance of the flywheel client
```
Comment (joshicola) [inline transfer_log.py:166]: There are 9 occurrences referencing `flywheel.Client` in this file.

2 are `fw`
4 are `fw_client`
3 are `client`

Do you want consistency with respects to their naming?
I like the `fw_client` above the legacy `fw` and to be just a little more descriptive than the `client`.

**Pull request merged**